compiler/rustc_infer/src/infer/error_reporting/need_type_info.rs

```diff
@@ -653,7 +654,41 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
             };
             err.span_label(pattern.span, msg);
         } else if let Some(e) = local_visitor.found_method_call {
-            if let ExprKind::MethodCall(segment, ..) = &e.kind {
+            if let ExprKind::MethodCall(segment, _, exprs, _) = &e.kind {
+                // Suggest impl candidates:
+                //
+                // error[E0283]: type annotations needed
+                //   --> $DIR/E0283.rs:35:24
+                //    |
+                // LL |     let bar = foo_impl.into() * 1u32;
+                //    |               ---------^^^^--
+                //    |               |        |
+                //    |               |        cannot infer type for type parameter `T` declared on the trait `Into`
+                //    |               this method call resolves to `T`
+                //    |               help: specify type like: `<Impl as Into<u32>>::into(foo_impl)`
+                //    |
+                //    = note: cannot satisfy `Impl: Into<_>`
+                if e.span.contains(span) {
+                    if let Some(expr) = exprs.first() {
+                        if let ExprKind::Path(hir::QPath::Resolved(_, path)) = expr.kind {
+                            if let Some(path_segment) = path.segments.first() {
+                                for impl_candidate in impl_candidates {
+                                    err.span_suggestion(
+                                        e.span,
+                                        "specify type like",
+                                        format!(
+                                            "{}::{}({})",
+                                            impl_candidate,
+                                            segment.ident.to_string(),
+                                            path_segment.ident.to_string()
+                                        ),
+                                        Applicability::MaybeIncorrect,
+                                    );
+                                }
+                            }
```

Comment:
A few nitpicks:

* let's make sure that there's a single path_segment and leverage the array pattern
* we already have a method for multiple exclusive suggestions, span_suggestions, let's use that one
* let's expand the suggestion message a little bit to use this as a chance to teach people the jargon
* anything that implements `to_string()` already implements `Display`, which means that you can use it directly in formatting strings

```suggestion
                            if let [path_segment] = &path.segments[..] {
                                let candidate_len = impl_candidates.len();
                                let suggestions: Vec<(Span, String)> = impl_candidates.iter().map(|candidate| {
                                    (e.span, format!("{}::{}({})", impl_candidate, segment.ident, path_segment.ident))
                                }).collect();
                                err.span_suggestions(
                                    &format!(
                                        "use the fully qualified path for the potential candidate{}",
                                        pluralize!(candidates_len),
                                    ),
                                    suggestions,
                                    Applicability::MaybeIncorrect,
                                );
                            }
```